### ISSUE TITLE
Unable to install Music Assistant on HASS supervised 

### ISSUE THREAD

**Issue by dinhchinh82** [music-assistant/hass-music-assistant#268]
Thank you for making the great music custom component for HASS. 
By the way, I'm running HASS on supervised environment and install Music Assistant 1.19 via HACS but not successfully.

Could you give me the suggestion how to fix.

Here is error log:

```
  2022-05-30 23:05:11 ERROR (SyncWorker_0) [homeassistant.util.package] Unable to install package music-assistant==1.1.19: error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [22 lines of output]
      running bdist_wheel
      running build
      running build_py
      running egg_info
      writing src/greenlet.egg-info/PKG-INFO
      writing dependency_links to src/greenlet.egg-info/dependency_links.txt
      writing requirements to src/greenlet.egg-info/requires.txt
      writing top-level names to src/greenlet.egg-info/top_level.txt
      reading manifest file 'src/greenlet.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      no previously-included directories found matching 'docs/_build'
      warning: no files found matching '*.py' under directory 'appveyor'
      warning: no previously-included files matching '*.pyc' found anywhere in distribution
      warning: no previously-included files matching '*.pyd' found anywhere in distribution
      warning: no previously-included files matching '*.so' found anywhere in distribution
      warning: no previously-included files matching '.coverage' found anywhere in distribution
      adding license file 'LICENSE'
      adding license file 'LICENSE.PSF'
      adding license file 'AUTHORS'
      running build_ext
      building 'greenlet._greenlet' extension
      error: command 'gcc' failed: No such file or directory
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for greenlet
  error: subprocess-exited-with-error
  
  × Running setup.py install for greenlet did not run successfully.
  │ exit code: 1
  ╰─> [24 lines of output]
      running install
      /usr/local/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_py
      running egg_info
      writing src/greenlet.egg-info/PKG-INFO
      writing dependency_links to src/greenlet.egg-info/dependency_links.txt
      writing requirements to src/greenlet.egg-info/requires.txt
      writing top-level names to src/greenlet.egg-info/top_level.txt
      reading manifest file 'src/greenlet.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      no previously-included directories found matching 'docs/_build'
      warning: no files found matching '*.py' under directory 'appveyor'
      warning: no previously-included files matching '*.pyc' found anywhere in distribution
      warning: no previously-included files matching '*.pyd' found anywhere in distribution
      warning: no previously-included files matching '*.so' found anywhere in distribution
      warning: no previously-included files matching '.coverage' found anywhere in distribution
      adding license file 'LICENSE'
      adding license file 'LICENSE.PSF'
      adding license file 'AUTHORS'
      running build_ext
      building 'greenlet._greenlet' extension
      error: command 'gcc' failed: No such file or directory
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> greenlet

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
WARNING: You are using pip version 22.0.3; however, version 22.1.1 is available.
You should consider upgrading via the '/usr/local/bin/python3 -m pip install --upgrade pip' command.
```

My environment:

```
Home Assistant Core 2022.5.5
Home Assistant Supervisor 2022.05.3
```

I also tried to install greenlet manually via pip3 install greenlet then try to install music assistant again but still same error.

Thank you.

**Comment by marcelveldt** [comment]
Strange, I've never seen this one before. What architecture are you running HA on ?
It looks like its trying to compile greenlet but that fails due to missing dev tools like gcc

**Comment by marcelveldt** [comment]
@dinhchinh82 if I understand correctly, you did manage to install greenlet manually from within the HA core container ?

see also #270 for a duplicate issue report

**Comment by dinhchinh82** [comment]
Here is the log from docker:
<img width="1348" alt="image" src="https://user-images.githubusercontent.com/22479095/171217816-ffb16b45-9f9a-4033-a3a6-757f77fc0c4d.png">

**Comment by dinhchinh82** [comment]
Here are more detail about my architecture:
<img width="644" alt="image" src="https://user-images.githubusercontent.com/22479095/171218668-db9ff001-3e38-4ddc-8484-8a0d606edab3.png">
<img width="606" alt="image" src="https://user-images.githubusercontent.com/22479095/171218748-afb68466-ac87-47d5-8581-6a6c2ef41625.png">

**Comment by dinhchinh82** [comment]
@marcelveldt I just tried to install 1.1.20 and still got the same error.

`2022-06-01 12:56:45 ERROR (SyncWorker_4) [homeassistant.util.package] Unable to install package music-assistant==1.1.20: error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [22 lines of output]
      running bdist_wheel
      running build
      running build_py
      running egg_info
      writing src/greenlet.egg-info/PKG-INFO
      writing dependency_links to src/greenlet.egg-info/dependency_links.txt
      writing requirements to src/greenlet.egg-info/requires.txt
      writing top-level names to src/greenlet.egg-info/top_level.txt
      reading manifest file 'src/greenlet.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      no previously-included directories found matching 'docs/_build'
      warning: no files found matching '*.py' under directory 'appveyor'
      warning: no previously-included files matching '*.pyc' found anywhere in distribution
      warning: no previously-included files matching '*.pyd' found anywhere in distribution
      warning: no previously-included files matching '*.so' found anywhere in distribution
      warning: no previously-included files matching '.coverage' found anywhere in distribution
      adding license file 'LICENSE'
      adding license file 'LICENSE.PSF'
      adding license file 'AUTHORS'
      running build_ext
      building 'greenlet._greenlet' extension
      error: command 'gcc' failed: No such file or directory
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for greenlet
  error: subprocess-exited-with-error
  
  × Running setup.py install for greenlet did not run successfully.
  │ exit code: 1
  ╰─> [24 lines of output]
      running install
      /usr/local/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_py
      running egg_info
      writing src/greenlet.egg-info/PKG-INFO
      writing dependency_links to src/greenlet.egg-info/dependency_links.txt
      writing requirements to src/greenlet.egg-info/requires.txt
      writing top-level names to src/greenlet.egg-info/top_level.txt
      reading manifest file 'src/greenlet.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      no previously-included directories found matching 'docs/_build'
      warning: no files found matching '*.py' under directory 'appveyor'
      warning: no previously-included files matching '*.pyc' found anywhere in distribution
      warning: no previously-included files matching '*.pyd' found anywhere in distribution
      warning: no previously-included files matching '*.so' found anywhere in distribution
      warning: no previously-included files matching '.coverage' found anywhere in distribution
      adding license file 'LICENSE'
      adding license file 'LICENSE.PSF'
      adding license file 'AUTHORS'
      running build_ext
      building 'greenlet._greenlet' extension
      error: command 'gcc' failed: No such file or directory
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> greenlet

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
WARNING: You are using pip version 22.0.3; however, version 22.1.2 is available.
You should consider upgrading via the '/usr/local/bin/python3 -m pip install --upgrade pip' command.`

**Comment by dinhchinh82** [comment]
I also tried to update pip3

`python3 -m pip install --upgrade pip
Requirement already satisfied: pip in /usr/local/lib/python3.9/dist-packages (22.1.1)
Collecting pip
  Downloading pip-22.1.2-py3-none-any.whl (2.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 1.2 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 22.1.1
    Uninstalling pip-22.1.1:
      Successfully uninstalled pip-22.1.1
Successfully installed pip-22.1.2`

and tried to install greenlet manually:

`pip3 install greenlet
Requirement already satisfied: greenlet in /usr/local/lib/python3.9/dist-packages (1.1.2)
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv`

**Comment by marcelveldt** [comment]
Question is why it is trying to install greenlet while that is already installed.
Version 1.1.20 of MA has greenlet pinned in the integration's manifest to hopefully solve this weird issue. Looks like that doesn't help and I'm out of options.

There's a good chance there's another (custom) integration or its dependencies or maybe some leftover cached stuff that is causing this issue but I'm not a pip expert to have the answer here. Unfortunately it is system specific because the majority of users can install just fine.

I will ask around today if somebody has any clue.
Maybe you can have a look at everything installed if you can spot something that might cause this.

**Comment by marcelveldt** [comment]
@dinhchinh82 did you try to install greenlet from within the HA core container itself or on the host ?

There's a little something to try... Go to your ha config folder --> custom_components/mass
edit the file manifest.json 
edit the "greenles==1.1.2" and change it to "greenlet==1.1.1" 

restart HA and try if it installs then.


Other than that I can't help you. You are running an unsupported config at the moment, for example armbian.
Not all packages/wheels are available. So without being rude I want to tell you that when you decide to create a supervised install with unsupported building blocks you're kind of on your own when it comes to issues.

If I were you I'd look into installing HA OS into your device, works rock stable for me.

**Comment by corgan2222** [comment]
I have a similar problem.

I installed this integration yesterday, which worked fine till today. 
But after I installed the new update today, I restarted HA and now get these errors:

> Setup failed for custom integration mass: Requirements for mass not found: ['greenlet==1.1.2'].

![grafik](https://user-images.githubusercontent.com/12233951/171381055-421e6f42-77a2-41d2-8796-4aeee77de9af.png)

**Comment by corgan2222** [comment]
> @corgan2222 see my note above about editing the manifest.json to see if that helps for you

This don't works for me

![grafik](https://user-images.githubusercontent.com/12233951/171389899-a7a4cf00-3d23-4a47-98e4-2918eb46db8d.png)

after restart:

![grafik](https://user-images.githubusercontent.com/12233951/171389861-118ce6c0-7fc3-417d-a776-9a194bff9e86.png)

**Comment by marcelveldt** [comment]
@corgan2222 and if you try to remove the requirement for greenlet altogether ?
or change it to plain 'greenlet' without the version ?

Are you also running supervised with the unsupported warning ?

**Comment by dinhchinh82** [comment]
@marcelveldt I found a strang point. I've tried to delete the mass version from HACS and also remove HACS then restart HASS. 
Later on, I download the latest source code from your repo and install manually to custom_component folder.
By the way, when I trying to add music assistant from Integration, the log still show the errore as before and I can see it is still trying to install version 1.1.20, not the latest version 1.1.22 as expected.

Please check the latest log:

> 2022-06-01 20:47:13 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 79, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 219, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 137, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 173, in post
    return await super().post(request)
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 62, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 70, in post
    result = await self._flow_mgr.async_init(
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 205, in async_init
    flow, result = await task
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 223, in _async_init
    flow = await self.async_create_flow(handler, context=context, data=data)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 740, in async_create_flow
    await async_process_deps_reqs(self.hass, self._hass_config, integration)
  File "/usr/src/homeassistant/homeassistant/setup.py", line 360, in async_process_deps_reqs
    await requirements.async_get_integration_with_requirements(
  File "/usr/src/homeassistant/homeassistant/requirements.py", line 83, in async_get_integration_with_requirements
    await _async_process_integration(hass, integration, done)
  File "/usr/src/homeassistant/homeassistant/requirements.py", line 99, in _async_process_integration
    await async_process_requirements(
  File "/usr/src/homeassistant/homeassistant/requirements.py", line 162, in async_process_requirements
    await _async_process_requirements(
  File "/usr/src/homeassistant/homeassistant/requirements.py", line 194, in _async_process_requirements
    raise RequirementsNotFound(name, [req])
homeassistant.requirements.RequirementsNotFound: Requirements for mass not found: ['music-assistant==1.1.20'].

**Comment by corgan2222** [comment]
> @corgan2222 and if you try to remove the requirement for greenlet altogether ? or change it to plain 'greenlet' without the version ?
> 
> Are you also running supervised with the unsupported warning ?

remove the version
![grafik](https://user-images.githubusercontent.com/12233951/171422020-0ceea2f8-be8a-4c23-86d9-a37d74e96c2f.png)

shows me this erros:
![grafik](https://user-images.githubusercontent.com/12233951/171422241-5643f862-b335-470a-bf12-f453b211d9fa.png)


but remove the line completely fixed the issue for me! Thanks!

**Comment by marcelveldt** [comment]
@corgan2222 thanks for testing. I've decided to ditch that pinned version of greenlet completely from the manifest file.
I have no idea what is going on with the few users that report issues with greenlet but I guess its simply not available as prebuilt wheel for their platform.

I am going to close this issue now because I can't fix the issue of unsupported platforms.

**Comment by dinhchinh82** [comment]
@marcelveldt I just installed latest version 1.1.22 which you already removed the greenlet but I confused why greenlet is still showing in the log.

<img width="1360" alt="image" src="https://user-images.githubusercontent.com/22479095/171441580-43da6348-fffa-4cec-bb1b-5bd3b642a898.png">

**Comment by marcelveldt** [comment]
@dinhchinh82 I did not remove the greenlet dependency. I just removed that pinned version from the manifest as it clearly did not fix your issues and instead caused other trouble. I do not have a solution for your issue I'm afraid, it's an issue with your set-up.

**Comment by marcelveldt** [comment]
Maybe you can try to build greenlet yourself and share it in the deps folder ?

**Comment by dinhchinh82** [comment]
> Maybe you can try to build greenlet yourself and share it in the deps folder ?

so you mean if I could build greenlet from scratch in my environment instead of using pre-built lib then it may work?

**Comment by sgueniot** [comment]
Hi, on my raspberry with HA core in docker I have a similar error. The problem (visible int the log : ` command 'gcc' failed: No such file or directory`) was that I needed to install gcc and some other dependencies in the docker container.
I am using portainer, I just opened the HA container console, it is based on alpine linux so I started the installation with 

```
apk update
apk add build-base
```
After the reboot, I restart the Music Assistant installation successfully.
I hope this helps someone
